lib/unifex/code_generator/base_types/list.ex

```diff
@@ -74,18 +81,100 @@ defmodule Unifex.CodeGenerator.BaseTypes.List do
     end
 
     @impl BaseType
-    def generate_initialization(name, _ctx) do
-      ~g<#{name} = NULL;>
+    def generate_destruction(name, ctx) do
+      ~g"""
+      if(#{name} != NULL) {
+        for(unsigned int i = 0; i < #{name}_length; i++) {
+          #{BaseType.generate_destruction(ctx.subtype, :"#{name}[i]", ctx.generator)}
+        }
+        unifex_free(#{name});
+      }
+      """
+    end
+  end
+
+  defmodule CNode do
+    @moduledoc false
+    use Unifex.CodeGenerator.BaseType
+    alias Unifex.CodeGenerator.BaseType
+    @impl BaseType
+    def generate_native_type(ctx) do
+      prefix = if ctx.mode == :const, do: "const ", else: ""
+
+      [
+        "#{prefix}#{BaseType.generate_native_type(ctx.subtype, ctx.generator)}*",
+        {"int", "_length"}
+      ]
+    end
+
+    @impl BaseType
+    def generate_arg_serialize(name, ctx) do
+      ~g"""
+      ({
+        ei_x_encode_list_header(out_buff, #{name}_length);
+        for(int i = 0; i < #{name}_length; i++) {
+          #{BaseType.generate_arg_serialize(ctx.subtype, :"#{name}[i]", ctx.generator)}
+        }
+        ei_x_encode_empty_list(out_buff);
+      });
+      """
+    end
+
+    @impl BaseType
+    def generate_arg_parse(arg, var_name, ctx) do
+      elem_name = :"#{var_name}[i]"
+      len_var_name = "#{var_name}_length"
+
+      ~g"""
+      ({
+        int res = 1;
+        int type;
+        ei_get_type(#{arg}->buff, #{arg}->index, &type, &#{len_var_name});
+        if(type == ERL_LIST_EXT) {
+          res = ei_decode_list_header(#{arg}->buff, #{arg}->index, &#{len_var_name});
+          #{var_name} = malloc(sizeof(#{BaseType.generate_native_type(ctx.subtype, ctx.generator)}) * #{len_var_name});
+
+          for(int i = 0; i < #{len_var_name}; i++) {
+            #{BaseType.generate_initialization(ctx.subtype, elem_name, ctx.generator)}
+          }
+
+          for(int i = 0; i < #{len_var_name}; i++) {
+            #{BaseType.generate_arg_parse(
+              ctx.subtype,
+              elem_name,
+              arg,
+              ctx.postproc_fun,
+              ctx.generator
+            )}
+          }
+        } else if(type == ERL_STRING_EXT) {
+          char *p = malloc(sizeof(char) * #{len_var_name});
+          res = ei_decode_string(#{arg}->buff, #{arg}->index, p);
+          #{var_name} = malloc(sizeof(int) * #{len_var_name});
+          for(int i = 0; i < #{len_var_name}; i++) {
+            #ifdef __CHAR_UNSIGNED
+              #{elem_name} = (int)p[i];
+            #else
+              #{elem_name} = (int)p[i];
+              if(#{elem_name} < 0) {
+                #{elem_name} = #{elem_name} + 256;
+              }
+            #endif
+            }
```

Comment:
This code is for list of ints between 0 and 255. Shouldn't we generate this only when `ctx.subtype` belongs to [int, unsigned int, int 64]`? It works but it's not too clear and compiler returns warnings if parse e.g. string. (cast warnings)

Reply:
There's one more problem there: this code doesn't return an error if someone accidentally passes such [0..255] list to a function accepting a list of strings, what may lead to a segfault. Generating the code only for some subtypes can lead to runtime errors if we add some can-be-integer subtype. What if we just converted such problematic lists to normal ones? We could decode them with ei_decode_string, then encode with `ei_x_encode_list_header` and `ei_x_encode_ulong` and finally decode normally.

Reply:
This sounds good and resolves problem with accidental segfaults but unfortunately it won't work. If we decode list using `decode_string` and then encode it using `encode_list_header` `encode_ulong` we still have to decode new list using `decode_string` :smile: 

here is code 
```c
        else if(type == ERL_STRING_EXT) {
          char *p = malloc(sizeof(char) * #{len_var_name});
          res = ei_decode_string(#{arg}->buff, #{arg}->index, p);

          ei_x_buff *buff = (ei_x_buff *)malloc(sizeof(ei_x_buff));
          ei_x_new_with_version(buff);
          ei_x_encode_list_header(buff, #{len_var_name});

          for(int i = 0; i < #{len_var_name}; i++) {
            int char_value;
            #ifdef __CHAR_UNSIGNED
              char_value = (int)p[i];
            #else
              char_value = (int)p[i];
              if(char_value < 0) {
                char_value = char_value + 256;
              }
            #endif
            printf("%d\n", char_value);
            fflush(stdout);
            ei_x_encode_ulong(buff, (unsigned int)char_value);
          }
          ei_x_encode_empty_list(buff);

          res = ei_decode_list_header(#{arg}->buff, #{arg}->index, &#{len_var_name});
          #{var_name} = malloc(sizeof(#{BaseType.generate_native_type(ctx.subtype, ctx.generator)}) * #{len_var_name});
          for (int i = 0; i < #{len_var_name}; i++) {
            unsigned long tmp_ulong;
            
            // can't do it in that way. have to use decode_string 
            res = ei_decode_ulong(#{arg}->buff, #{arg}->index, &tmp_ulong); here. 
            printf("dec: %d\n", tmp_ulong);
            fflush(stdout);
            #{elem_name} = (#{BaseType.generate_native_type(ctx.subtype, ctx.generator)})tmp_ulong;
          }
        }
        res;
      })
```

Reply:
Maybe I should just do sth like this
```c
        for(int i = 0; i < #{len_var_name}; i++) {
            #ifdef __CHAR_UNSIGNED
              #{elem_name} = (#{BaseType.generate_native_type(ctx.subtype, ctx.generator)})p[i];
            #else
              int tmp = (int)p[i]
              if(tmp < 0) {
                tmp = tmp + 256;
              }
              #{elem_name} = (#{BaseType.generate_native_type(ctx.subtype, ctx.generator)})tmp;
            #endif
            }
```
`(#{BaseType.generate_native_type(ctx.subtype, ctx.generator)})tmp` will remove compiler warnings.